heat_pump_adoption_modelling/pipeline/MCS/mcs_epc_joining.py

```diff
@@ -0,0 +1,322 @@
+# File: heat_pump_adoption_modelling/pipeline/MCS/mcs_epc_joining.py
+"""Joining the MCS and EPC datasets."""
+
+import pandas as pd
+import numpy as np
+import string
+import re
+import recordlinkage as rl
+import time
+
+from heat_pump_adoption_modelling import PROJECT_DIR
+from heat_pump_adoption_modelling.pipeline.MCS.load_mcs import (
+    load_domestic_hps,
+    load_epcs,
+)
+
+
+matching_parameter = 0.7
+merged_path = "outputs/mcs_epc.csv"
+# TODO: put this in config
+
+
+# Overall process:
+# - standardise address and postcode fields
+# - extract numeric tokens from address
+# - group by postcode
+# - exact match on numeric tokens
+# - compare address using jaro-winkler
+# - drop anything below a certain parameter
+# - of remaining matches, take the best
+# - join using this matching
+
+#### UTILS
+
+
+def rm_punct(address):
+    """Remove all unwanted punctuation from an address.
+    Underscores are kept and slashes/dashes are converted
+    to underscores so that the numeric tokens in e.g.
+    "Flat 3/2" and "Flat 3-2" are treated as a whole later.
+
+    Parameters
+    ----------
+    address : str
+        Address to format.
+
+    Return
+    ----------
+    address : str
+        Formatted address."""
+    if (address is pd.NA) | (address is np.nan):
+        return ""
+    else:
+        # replace / and - with _
+        address = address.replace("/", "_").replace("-", "_")
+        # remove all punctuation other than _
+        address = address.translate(
+            str.maketrans("", "", string.punctuation.replace("_", ""))
+        )
+        return address
+
+
+def extract_numeric_tokens(df):
```

Comment:
I'd recommend adding functionality for removing extra long numeric tokens. They tend to not match anything in EPC anyways. Here some examples:

![Screenshot 2021-11-03 at 17 40 55](https://user-images.githubusercontent.com/42718928/140320078-f61d0c09-f946-456a-9467-9ead673ce59e.png)

You may also want to remove numeric tokens that match the postcode for that entry.  For example:
![Screenshot 2021-11-03 at 23 56 37](https://user-images.githubusercontent.com/42718928/140320175-0d264180-3624-4a24-ad0b-b4317be52571.png)

Reply:
Great idea - have done this